debezium-core/src/main/java/io/debezium/pipeline/signal/channels/process/InProcessSignalChannel.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.signal.channels.process;
+
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.DebeziumException;
+import io.debezium.annotation.ThreadSafe;
+import io.debezium.config.CommonConnectorConfig;
+import io.debezium.pipeline.signal.SignalRecord;
+import io.debezium.pipeline.signal.channels.SignalChannelReader;
+
+/**
+ * Implementation of {@link SignalChannelReader} that also implements {@link SignalChannelWriter}
+ * used for sending signal from the same JVM process
+ *
+ * <p>
+ *     Mainly targeted at Debezium Engine
+ * </p>
+ */
+@ThreadSafe
+public class InProcessSignalChannel implements SignalChannelReader, SignalChannelWriter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InProcessSignalChannel.class);
+    public static final String CHANNEL_NAME = "in-process";
+
+    private final AtomicBoolean open = new AtomicBoolean(false);
+    private final Queue<SignalRecord> signals = new ConcurrentLinkedQueue<>();
+
+    @Override
+    public String name() {
+        return CHANNEL_NAME;
+    }
+
+    @Override
+    public void init(CommonConnectorConfig connectorConfig) {
+        open.compareAndSet(false, true);
+        LOGGER.info("Reading signals from {} channel", CHANNEL_NAME);
+    }
+
+    @Override
+    public List<SignalRecord> read() {
+        return Stream.ofNullable(signals.poll()).toList();
```

Comment:
possibly also add check if the channel is open here?

Reply:
Is there a point in doing that? I mean there is no harm but in case it isn't the read will just return empty list.  Which TBF looks safer (there is no danger of exception) in case we overlooked some concurrency sync point and read is called when not expected (no need to be extra careful here due to the nature of what signals are).

Reply:
Actually it's safer to throw an exception, in case you overlook something, you are notified you overlooked something (and there may be some more side effects). Also, when you read from closed channel and wait for signal, you would be stuck there forever. Basically reading from closed channel shouldn't happen, if it happen it should be either fixed (better option) or the exception can be used as a signal to closer the signal reader (not that great option to use exception as signals, but IMHO still better than reading from closed channel without noticing it's closed).

Reply:
> you would be stuck there forever

That's not true, all our channels return an empty List in such case. This would be the only one which throws an exception

The "close" here could very well be a no-op here. As there really isn't any resource which needs closing. It's implemented this way only to notify the sending side that the connector is shutting down and thus the signal would not be processed.

However, thinking about it there might be a bit of an issue when close is draining the queue and read is called, so I will add the check ;)

Reply:
Well, if all our signaler do so (which I checked and it's correct), let's keep it this way. That being said, I still consider reading from a closed resource as an error which shouldn't happen and throw an exception in such case.